lib/admin/components/ServerSettings.js

```diff
@@ -0,0 +1,269 @@
+// @flow
+
+import Icon from '@conveyal/woonerf/components/icon'
+import objectPath from 'object-path'
+import React, {Component} from 'react'
+import {connect} from 'react-redux'
+import Select from 'react-select'
+import {
+  Row,
+  Col,
+  Button,
+  Panel,
+  Glyphicon,
+  FormGroup,
+  ControlLabel
+} from 'react-bootstrap'
+import update from 'react-addons-update'
+
+import { getComponentMessages } from '../../common/util/config'
+import * as adminActions from '../actions/admin'
+import CollapsiblePanel from '../../manager/components/CollapsiblePanel'
+import { FIELDS, SERVER_FIELDS, UPDATER_FIELDS } from '../../manager/util/deployment'
+
+import type { OtpServer, Project } from '../../types'
+import type { AppState, AdminServersState } from '../../types/reducers'
+
+type ContainerProps = {editDisabled: boolean}
+type Props = ContainerProps & {
+  deleteServer: typeof adminActions.deleteServer,
+  otpServers: AdminServersState,
+  projects: Array<Project>,
+  updateServer: typeof adminActions.updateServer
+}
+
+type State = {
+  otpServers: Array<OtpServer>
+}
+
+class ServerSettings extends Component<Props, State> {
+  messages = getComponentMessages('ServerSettings')
+
+  state = {
+    otpServers: []
+  }
+
+  componentWillMount () {
+    this._updateStateFromProps(this.props)
+  }
+
+  componentWillReceiveProps (nextProps) {
+    this._updateStateFromProps(nextProps)
+  }
+
+  _updateStateFromProps = props => {
+    if (props.otpServers && props.otpServers.data) {
+      this.setState({ otpServers: props.otpServers.data })
+    }
+  }
+
+  _getOnChange = (evt, index) => {
```

Comment:
This is beginning to look like we're inventing a React form framework. This isn't a good place for that because it's within just this component. We should not have every single component have their own form frameworks and instead just pick one and use it everywhere to be consistent. I liked using [React-Formal](http://jquense.github.io/react-formal/) in the commute project.

Reply:
Good point. Much of this form handling code was taken out of `DeploymentSettings`. Do you think we should create a separate issue/PR for discussing/implementing some other form framework?

Reply:
Yeah, would be good to create an issue for that.